spec/controllers/password_resets_controller_spec.rb

```diff
@@ -0,0 +1,250 @@
+      end
+
+      it 'creates a reset digest for user' do
+        expect(user).to have_received(:create_reset_digest)
+      end
+
+      it 'sends an email to user' do
+        expect(user).to have_received(:send_password_reset_email)
+      end
+
+      it 'sets flash warning message' do
+        expect(flash[:warning]).to be_present
+      end
+
+      it 'redirects to the root page' do
+        expect(response).to redirect_to(root_url)
+      end
+
+      it 'responds with redirect status' do
+        expect(response).to have_http_status(:redirect)
+      end
+    end
+
+    context 'with unactivated user' do
+      before do
+        allow(user).to receive(:activated?).and_return(false)
+        post :create, params: { password_reset: { email: email } }
+      end
+
+      it "doesn't create a reset digest for user" do
+        expect(user).to_not have_received(:create_reset_digest)
+      end
+
+      it "doesn't send an email to user" do
+        expect(user).to_not have_received(:send_password_reset_email)
+      end
+
+      it 'sets flash warning message' do
+        expect(flash[:warning]).to be_present
+      end
+
+      it 'redirects to the root page' do
+        expect(response).to redirect_to(root_url)
+      end
+
+      it 'responds with redirect status' do
+        expect(response).to have_http_status(:redirect)
+      end
+    end
+  end
+
+  shared_examples 'invalid user' do
+    it 'sets flash alert message' do
+      expect(flash[:alert]).to be_present
+    end
+
+    it 'redirects to the root page' do
+      expect(response).to redirect_to(root_url)
+    end
+
+    it 'responds with redirect status' do
+      expect(response).to have_http_status(:redirect)
+    end
+  end
+
+  shared_examples 'expired reset token' do
+    it 'sets flash alert message' do
+      expect(flash[:alert]).to be_present
+    end
+
+    it 'redirects to new reset password page' do
+      expect(response).to redirect_to(new_password_reset_url)
+    end
+
+    it 'responds with redirect status' do
+      expect(response).to have_http_status(:redirect)
+    end
+  end
+
+  describe 'GET #edit' do
+    let(:user) { create(:user) }
+    let(:email) { user.email }
+
+    before do
+      allow(User).to receive(:find_by).with(email: email).and_return(user)
+    end
+
+    context 'valid user and with token not expired' do
+      before do
+        allow(controller).to receive(:valid_user)
+        allow(controller).to receive(:check_expiration)
+        get :edit, params: { id: user, email: email }
+      end
+
+      it 'renders edit' do
+        expect(response).to render_template(:edit)
+      end
+
+      it 'responds with success status' do
+        expect(response).to have_http_status(:success)
+      end
+    end
+
+    context 'invalid user' do
+      before do
+        allow(user).to receive(:activated?).and_return(true)
+        allow(user).to receive(:authentic_token?).with(:reset, user.id.to_s)
```

Comment:
Either `:activated?` or `:authentic_token?` should return false to reproduce the invalid user flow. Actually, one of them at least.